### DISCUSSION TITLE
Package auch ohne node red nutzbar? (Version 1.x)

### DISCUSSION THREAD

**Discussion by kainers1** [vrilcode/node-red-wnsm#1] [General]
Es es möglich dieses Package auch ohne node red zu nutzen?
Wenn ja, würde ich mich über ein kurzes nodejs-Beispiel freuen.
Danke!

**Comment by vrilcode** [comment]
Ja, du kannst die Klasse `WnsmB2b` auch direkt verwenden. Einfach das npm package installieren und direkt einbinden oder die Datei `wnsm-b2b.js` ins eigene Projekt kopieren. Hier ein Minimalbeispiel mit npm package:

```js
const WnsmB2b = require ('@vrilcode/node-red-wnsm-b2b/wnsm-b2b')

const username = 'user@example.com'
const password = 'password123'

const wnsm = new WnsmB2b(username, password)

const dateFrom = '2023-10-01'
const dateTo = '2023-10-31'

wnsm.call(
  '/zaehlpunkte/messwerte/',
  {
    'datumVon': dateFrom,
    'datumBis': dateTo,
    'wertetyp': 'METER_READ'
  }
).then(res => {
  console.log(JSON.stringify(res, null, 2))
}).catch(err => {
  console.error(err)
})
```

Welche weiteren REST-Endpoints von Wiener Netze bereitgestellt werden findest du [hier](https://api-portal.wienerstadtwerke.at/portal/apis/7f8a1cce-2a7e-4b18-840b-b0387ed9a3fc/apiinfo) unter "API resources" bzw. im Quelltext [hier](https://github.com/vrilcode/node-red-wnsm-b2b/blob/main/wnsm-b2b-node.js).

**Comment by kainers1** [comment]
Vielen Dank für das Beispiel, klappt perfekt.
Hast du evtl. auch die Anbindung der B2C-Api in Planung oder eher nicht? Da das Package 'b2b' heißt, würde ich mal letzteres denken.

**Reply by vrilcode** [reply]
Gerne. Ist derzeit nicht geplant. Der Code ist für einen Kunden, der die Zählerwerte zur Weiterverrechnung benötigt. Das sollte halbwegs stabil laufen. Der derzeitige Login-Hack soll daher künftig auch irgendwann ausgebaut und auf eine reguläre API-Authentifizierung umgestellt werden. Dann muss man nicht ständig nachbessern, falls die Wiener Netze / Wiener Stadtwerke irgendwas beim Login verändern. Soweit ich gesehen habe, ist die offizielle API + Authentifizierung nur für B2B geplant (wieso auch immer). Für B2C sind offiziell nur CSV-Downloads vorgesehen.

**Comment by kainers1** [comment]
WN dürfte was an der API geändert haben, musste den endpoint auf `/zaehlpunkte/12345678/AT00000000/messwerte/` ändern. Bei WN ist die alte Variante aber noch in der [Api-Definition](https://api-portal.wienerstadtwerke.at/portal/apis/7f8a1cce-2a7e-4b18-840b-b0387ed9a3fc/resources/4d6d5db0-4472-4a24-b417-66d819d3c636/GET).
Weiß nicht ob das für die node red-Implementierung auch eine Auswirkung hat, habe gesehen dass im Code teilweise Kundennummer und Zählpunkt verwendet wird.

> Der derzeitige Login-Hack soll daher künftig auch irgendwann ausgebaut und auf eine reguläre API-Authentifizierung umgestellt werden.

Bekommt man als Privatkunde einen solchen api-key? Meine lt. WN-Anleitung eingereichte Applikation wurde abgelehnt.

**Reply by vrilcode** [reply]
Ad "Änderung": Grds. sollte es laut Doku beide Varianten geben. Die Variante, wo man alle Zählpunkte, die dem Benutzerkonto zugeordnet sind, auf einmal abrufen kann, liefert bei mir derzeit HTTP 500. Ob diese Änderung irrtümlich oder absichtlich passiert ist, ist schwer zu sagen, weil die WN-API schnell mal HTTP 500 liefert. Ich gehe aber davon aus, dass man das wieder funktionstüchtig machen wird, zumal diese Variante auch ausführlich im API-PDF beschrieben ist und inhaltlich sinnvoll wäre.

Ad "Implementierung": Auswirkungen auf die Node-RED-node hat es wenig, weil man diese so einstellen kann, dass eine Liste aller Zählpunkte abgerufen wird. Auf Basis dieser Liste kann man dann die einzelnen Zählpunkte mit einer zweiten node abrufen. Es ist also eine Spur umständlicher als bisher, wenn man die Messwerte aller Zählpunkte möchte, aber es ist möglich mit dieser Implementierung. Ich lasse die Funktion für den Abruf aller Zählpunkt-Messwerte erst einmal drin, weil ich davon ausgehe, dass WN das wieder fixen wird.

Ad "B2C": Ich habe nirgendwo in den Unterlagen/Anleitungen oder auf der Weboberfläche gesehen, dass die Bereitstellung der API für B2C geplant ist. Ich glaube man will dem Privatkunden nur den CSV-Datenabruf ermöglichen (wieso auch immer). Aber vielleicht ändert sich das noch. Ich bin mir auch nicht sicher, ob die überhaupt in den Daten hinterlegt haben, ob es ein B2C- oder B2B-Kunde ist. Hast du schonmal versucht dich mit einem "B2C-Account" im B2B-Portal anzumelden? Ich habe nur "B2B-Accounts" und kann das nicht testen (umgekehrt komme ich jedenfalls auch in das B2C-Portal, was ich aber nicht brauche). Es würde mich nicht wundern, wenn man *derzeit* auch als B2B-Kunde keinen Zugriff auf die API bekommt. Das was man da beantragen kann, scheint nicht das zu sein, was am Schluss bereitgestellt werden soll. Im B2B-Portal sieht man, dass eine Antrag zur Freischaltung der API über das B2B-Portal geplant ist (siehe Screenshot nachstehend). Bei der großen Zahl an B2B-Kunden, scheint mir jede Freischaltung per E-Mail-Antrag o.ä. auch als Überforderung der ohnehin in allen Belangen überforderten Wiener Netze.

Da die Schnittstelle immer wieder ohne Vorwarnung geändert wird und ihre Macken hat (HTTP 500) schließe ich daraus, dass man derzeit noch daran entwickelt.

![image](https://github.com/vrilcode/node-red-wnsm-b2b/assets/31141078/cbc862c3-e8d8-4981-8c1e-e5d7f049bb14)

**Reply by kainers1** [reply]
Hier ist noch eine Rückmeldung von mir offen.

Betreffend Api-Zugang für B2C: Mit besserem Namen wurde inzwischen meine Applikation im Api-Portal genehmigt und ich habe die Email an die support.sm-portal Adresse geschickt. Empfang wurde bestätigt aber Auftrag noch nicht durchgeführt weil das extra gemacht wird.
Ins B2B-Portal komme ich ganz normal rein, habe halt dort nur einen Zähler zur Auswahl. Funktionen usw sind alle verwendbar.

Ich muss jetzt ohnehin mal die Freischaltung abwarten.

**Reply by kainers1** [reply]
Zwischenzeitlich wurde der Auftrag erledigt und ich habe im Kundenportal client_id und client_secret. Mit dem api-key aus dem Api-Portal kann ich in Postman erfolgreich Requests absetzen.

[Hier](https://github.com/tschoerk/Wiener-Netze-Smart-Meter-API/) habe ich noch eine Implementierung in Python gefunden.

**Reply by vrilcode** [reply]
Ich habe das auch schon probiert und es funktioniert. Die Implementierung dieser Schnittstelle als Node-RED-Node ist von mir geplant und es wird ein neues Projekt sein, da es nicht - wie ursprünglich von WN kommuniziert - allein auf B2B eingeschränkt ist. Ich hoffe, ich komme bald dazu.

**Reply by vrilcode** [reply]
Die neue Version wird vermutlich diese oder nächste Woche bereitstehen.

**Reply by vrilcode** [reply]
Die neue Version 2.0 ist ab sofort verfügbar. Achtung: Der Name hat sich geändert. Für neue Themen bitte einen neuen Thread aufmachen.
